cmd/outline-ss-server/main.go

```
@@ -340,48 +345,37 @@ func (s *OutlineServer) runConfig(config Config) (func() error, error) {
 							return fmt.Errorf("websocket-stream listener references unknown web server `%s`", cfg.WebsocketStream.WebServer)
 						}
 						mux := webServers[cfg.WebsocketStream.WebServer]
-						// TODO: Support a "half-closed" state for WebSockets.
 						handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-							handler := func(wsConn *websocket.Conn) {
-								defer wsConn.Close()
-								ctx, contextCancel := context.WithCancel(context.Background())
-								defer contextCancel()
-								clientIP, err := onet.GetClientIPFromRequest(r)
-								if err != nil {
-									slog.Error("failed to determine client address", "err", err)
-									w.WriteHeader(http.StatusBadGateway)
-									return
-								}
-								conn := &streamConn{&replaceAddrConn{Conn: wsConn, raddr: &net.TCPAddr{IP: clientIP}}}
-								streamHandler.HandleStream(ctx, conn, s.serviceMetrics.AddOpenTCPConnection(conn))
+							wsConn, err := defaultUpgrader.Upgrade(w, r, nil)
+							if err != nil {
+								slog.Error("failed to upgrade", "err", err)
 							}
-							websocket.Handler(handler).ServeHTTP(w, r)
+							defer wsConn.Close()
+							ctx, contextCancel := context.WithCancel(context.Background())
+							defer contextCancel()
+							conn := &replaceAddrConn{StreamConn: websocket.WrapConn(wsConn), raddr: &net.TCPAddr{IP: net.ParseIP(r.RemoteAddr)}}
```

I'm not sure how this could happen, but if ParseIP fails (absent remote address?), you will likely hit a null pointer exception down the line

---

Is that even possible? Let me be more defensive here though. Let's just throw an error in that case.